### PR TITLE
refactor(NODE-6201): cursor to use fetchBatch when current batch is empty

### PR DESCRIPTION
**Pull request by nbbeeken**
#### What is changing?

- Move `next` into cursor and rename to fetchBatch
- Remove transform and blocking flags
- tryNext/next now call fetchBatch as appropriate to get the same behavior the flags provided
- move cleanupCursor logic to cleanup()

##### Is there new documentation needed for these changes?

No

#### What is the motivation for this change?

The `next` floating function owned too many details of the implementation of each cursor API, tryNext and next now do a small amount more lifting and we have a fetchBatch method that is focused on obtaining the next batch either with the init command or with a getMore.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/sessions.ts [hunk 1]
```diff
@@ -523,7 +523,11 @@ export function maybeClearPinnedConnection(
     const servers = Array.from(topology.s.servers.values());
     const loadBalancer = servers[0];
 
-    if (options?.error == null || options?.force) {
+    if (
+      options?.error == null ||
+      options?.error?.name === 'MongoExpiredSessionError' ||
+      options?.force
+    ) {
```
Comment (nbbeeken) [inline src/sessions.ts:530]: Without this change, endSession is sensitive to the exact order of endSession calls that were made by our cursor before this refactor.

File: src/cursor/abstract_cursor.ts [hunk 2]
```diff
@@ -309,33 +305,26 @@ export abstract class AbstractCursor<
...
+          return;
+        }
+
+        if (this.closed && this[kDocuments].length === 0) {
+          return;
+        }
+
+        if (this[kId] != null && this.isDead && this[kDocuments].length === 0) {
+          return;
+        }
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts:318]: Yes, and ty! I intended these to be written as if you would read them:
- if killed - return
- if closed AND there's no more documents (because you can be closed but still finishing the last batch) - return
- if there is a truthy Id (which means we already ran the initial command) the cursor isDead (generally, id = 0) and there are no more documents - return

File: src/cursor/abstract_cursor.ts [hunk 3]
```diff
@@ -137,13 +133,13 @@ export abstract class AbstractCursor<
   CursorEvents extends AbstractCursorEvents = AbstractCursorEvents
 > extends TypedEventEmitter<CursorEvents> {
   /** @internal */
-  [kId]: Long | null;
+  private [kId]: Long | null;
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts:136]: Nope, I am happy to do that in another PR if this one is otherwise good to go, wouldn't take long

File: src/cursor/abstract_cursor.ts [hunk 4]
```diff
@@ -691,243 +695,156 @@ export abstract class AbstractCursor<
...
-      if (!session.inTransaction()) {
-        maybeClearPinnedConnection(session, { error });
+  /** @internal */
+  private hasEmittedClose = false;
+  /** @internal */
+  private emitClose() {
+    try {
+      if (!this.hasEmittedClose && (this[kDocuments].length === 0 || this[kClosed])) {
+        // @ts-expect-error: not sure..
+        this.emit('close');
```
Comment (baileympearson) [inline src/cursor/abstract_cursor.ts:807]: If we're going to leave a ts-expect-error in, can we document why it's an error?  but tbh I'd expect this to work, not sure why an expect error is needed here.

**Review by W-A-James — APPROVED** [review]

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: src/cursor/abstract_cursor.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts:136]: #4133

File: test/integration/node-specific/cursor_async_iterator.test.js [hunk 5]
```diff
@@ -72,6 +72,18 @@ describe('Cursor Async Iterator Tests', function () {
       }
     });
 
+    it('should not iterate if closed immediately', async function () {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/integration/node-specific/cursor_async_iterator.test.js:75]: From slack, we're showing "no documents are produced" here, a small edge case but proves we don't throw nor return `{value: null, done: false}` from our async iterator.

File: src/cursor/abstract_cursor.ts (hunk 4 above)
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts]: Yes sorry meant to get back to this is is declared in the same way ConnectionEvents are so I'm not sure what is going on here.

```ts
Argument of type '[]' is not assignable to parameter of type 'Parameters<CursorEvents["close"]>'.ts(2345)
```

Maybe it is because CursorEvents is generic so Parameters<CursorEvents["close"]> may not be assignable to `[]`. Not sure how to require extenders do not add parameters.

File: src/cursor/abstract_cursor.ts [hunk 6]
```diff
@@ -515,14 +529,14 @@ export abstract class AbstractCursor<
    * @param transform - The mapping transformation method.
    */
   map<T = any>(transform: (doc: TSchema) => T): AbstractCursor<T> {
-    assertUninitialized(this);
+    this.throwIfInitialized();
     const oldTransform = this[kTransform] as (doc: TSchema) => TSchema; // TODO(NODE-3283): Improve transform typing
     if (oldTransform) {
       this[kTransform] = doc => {
         return transform(oldTransform(doc));
       };
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts]: Good catch, added a test for this too

File: src/cursor/abstract_cursor.ts [hunk 7]
```diff
@@ -691,243 +695,156 @@ export abstract class AbstractCursor<
...
+          const message =
+            'Cursor returned a `null` document, but the cursor is not exhausted.  Mapping documents to `null` is not supported in the cursor transform.';
+          throw new MongoAPIError(message);
+        }
 
-  if (session.hasEnded) {
-    return await completeCleanup();
+        return result;
+      } catch (error) {
+        // eslint-disable-next-line github/no-then
```
Comment (baileympearson) [inline src/cursor/abstract_cursor.ts:828]: (optional) it might make more sense to propagate errors from `transform` and instead try-catch the application of the transform each place we use it (which, I think is just one place? `next()`?)

It seems like a small code smell to wrap a synchronous function in another synchronous function that sometimes kicks off an async operation but doesn't await it, when we can remove the async portion of the transform wrapper by calling await cursor.close() in the caller, not the transform.

**Review by baileympearson — DISMISSED** [review]
One optional comment.

File: src/cursor/abstract_cursor.ts (hunk 7 above)
Comment (baileympearson) [reply src/cursor/abstract_cursor.ts:828]: Or there's no reason `makeSafeTransform` can't be async, since we await it everywhere we use it anyways.

**Review by W-A-James — DISMISSED** [review]

File: src/cursor/abstract_cursor.ts (hunk 7 above)
Comment (nbbeeken) [reply src/cursor/abstract_cursor.ts]: `transformDocument` now awaits the close! and I removed the makeTransform from stream because we didn't disallow null / try/catch previously

**Review by baileympearson — APPROVED** [review]

**Review by W-A-James — APPROVED** [review]
